File: javav2/example_code/sns/src/main/java/com/example/sns/SubscribeTextSMS.java
//snippet-sourcedescription:[SubscribeTextSMS.java demonstrates how to subscribe to an Amazon Simple Notification Service (Amazon SNS) text endpoint.]
//snippet-keyword:[AWS SDK for Java v2]
//snippet-keyword:[Amazon Simple Notification Service]

/*
   Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
   SPDX-License-Identifier: Apache-2.0
*/

package com.example.sns;

//snippet-start:[sns.java2.SubscribeTextSMS.import]
import software.amazon.awssdk.auth.credentials.ProfileCredentialsProvider;
import software.amazon.awssdk.regions.Region;
import software.amazon.awssdk.services.sns.SnsClient;
import software.amazon.awssdk.services.sns.model.SnsException;
import software.amazon.awssdk.services.sns.model.SubscribeRequest;
import software.amazon.awssdk.services.sns.model.SubscribeResponse;
//snippet-end:[sns.java2.SubscribeTextSMS.import]

/**
 * Before running this Java V2 code example, set up your development environment, including your credentials.
 *
 * For more information, see the following documentation topic:
 *
 * https://docs.aws.amazon.com/sdk-for-java/latest/developer-guide/get-started.html
 */
public class SubscribeTextSMS {

    public static void main(String[] args) {

        final String usage = "\n" +
            "Usage: " +
            "   <topicArn> <phoneNumber>\n\n" +
            "Where:\n" +
            "   topicArn - The ARN of the topic to subscribe.\n\n" +
            "   phoneNumber - A mobile phone number that receives notifications (for example, +1XXX5550100).\n\n";

        if (args.length < 2) {
            System.out.println(usage);
            System.exit(1);
        }

        String topicArn = args[0];
        String phoneNumber = args[1];
        SnsClient snsClient = SnsClient.builder()
            .region(Region.US_EAST_1)
            .credentialsProvider(ProfileCredentialsProvider.create())
            .build();

        subTextSNS(snsClient, topicArn, phoneNumber);
        snsClient.close();
    }

    //snippet-start:[sns.java2.SubscribeTextSMS.main]
    public static void subTextSNS( SnsClient snsClient, String topicArn, String phoneNumber) {

        try {
            SubscribeRequest request = SubscribeRequest.builder()
                .protocol("sms")
                .endpoint(phoneNumber)
                .returnSubscriptionArn(true)
                .topicArn(topicArn)
                .build();

            SubscribeResponse result = snsClient.subscribe(request);
            System.out.println("Subscription ARN: " + result.subscriptionArn() + "\n\n Status is " + result.sdkHttpResponse().statusCode());

        } catch (SnsException e) {
            System.err.println(e.awsErrorDetails().errorMessage());
            System.exit(1);
        }
    }
    //snippet-end:[sns.java2.SubscribeTextSMS.main]
}

File: kotlin/services/sns/src/main/kotlin/com/kotlin/sns/SubscribeTextSMS.kt
// snippet-sourcedescription:[SubscribeTextSMS.kt demonstrates how to subscribe to an Amazon Simple Notification Service (Amazon SNS) text endpoint.]
// snippet-keyword:[AWS SDK for Kotlin]
// snippet-keyword:[Amazon Simple Notification Service]

/*
   Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
   SPDX-License-Identifier: Apache-2.0
*/

package com.kotlin.sns

// snippet-start:[sns.kotlin.SubscribeTextSMS.import]
import aws.sdk.kotlin.services.sns.SnsClient
import aws.sdk.kotlin.services.sns.model.SubscribeRequest
import kotlin.system.exitProcess
// snippet-end:[sns.kotlin.SubscribeTextSMS.import]

/**
Before running this Kotlin code example, set up your development environment,
including your credentials.

For more information, see the following documentation topic:
https://docs.aws.amazon.com/sdk-for-kotlin/latest/developer-guide/setup.html
 */
suspend fun main(args: Array<String>) {

    val usage = """
        
        Usage: 
            <topicArn> <phoneNumber>

        Where:
            topicArn - The ARN of the topic to publish.
            phoneNumber - A mobile phone number that receives notifications (for example, +1XXX5550100).
            """

    if (args.size != 2) {
        println(usage)
        exitProcess(0)
    }

    val topicArn = args[0]
    val phoneNumber = args[1]
    subTextSNS(topicArn, phoneNumber)
}

// snippet-start:[sns.kotlin.SubscribeTextSMS.main]
suspend fun subTextSNS(topicArnVal: String?, phoneNumber: String?) {

    val request = SubscribeRequest {
        protocol = "sms"
        endpoint = phoneNumber
        returnSubscriptionArn = true
        topicArn = topicArnVal
    }

    SnsClient { region = "us-east-1" }.use { snsClient ->
        val result = snsClient.subscribe(request)
        println("The subscription Arn is ${result.subscriptionArn}")
    }
}
// snippet-end:[sns.kotlin.SubscribeTextSMS.main]
